### PR TITLE
HDDS-8115. Do not use block count to determine if containers are empty.

### PR DESCRIPTION
**Pull request by ashishkumar50**
## What changes were proposed in this pull request?

Ozone currently uses a block count metadata field for each container replica to determine whether it is empty. This counter can go wrong sometime and also difficult to maintain.

In this PR we simplify check for container flow, which will now make use of isEmpty boolean flag to determine whether container replica is empty or not. When we create container, isEmpty flag starts as false and is set to true when a delete block is processed and the block table is empty. This will ease in making decision for checking empty container and later we can delete those empty container. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8115

## How was this patch tested?

Added new test case as well as verified using existing test case.

### CODE REVIEW

**Comment by ashishkumar50** [comment]
@errose28 , Can you please help to review.

**Comment by kerneltime** [comment]
Please resolve the conflict

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java
Comment (Xushaohong) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:410]: Why the container is considered to be empty when it is not open?  It is common that the container has blocks when it transitions into states like  CLOSING.

Comment (ashishkumar50) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:410]: When container is open it is considered as not empty. When in any other state(other than open state) we are checking in block table to see if there are any block present and then we decide container is empty or not.

**Comment by ashishkumar50** [comment]
> Please resolve the conflict

Resolved conflict.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java
Comment (devmadhuu) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:412]: This PR is being done and added a flag due to inconsistency in block count data, so what brings inconsistency and why that inconsistency will not impact above logic ?

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationTestUtil.java
Comment (devmadhuu) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationTestUtil.java:163]: Should this be keyCount or blockCount ?

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/HddsTestUtils.java
Comment (devmadhuu) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/HddsTestUtils.java:739]: Should this be keyCount or blockCount ?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java
Comment (ashishkumar50) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:410]: Yes keeping block count metadata can go wrong sometime which leads to wrong interpretation of empty container. 
Now we check if any real block exists in container(not counting all block count), Which is done in isEmpty() method.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java
Comment (ashishkumar50) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:412]: We keep block count metadata information separately about container, which can go wrong sometime due to various reasons. Now we decouple metadata information to judge if container is empty or not by checking directly block table if any block exists or not for the particular container.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/HddsTestUtils.java
Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/HddsTestUtils.java:739]: This is in server-scm which doesn't use block count. Here existing test case is just to determine whether replica has some keys or not. If some keys are present then setEmpty as false.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationTestUtil.java
Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationTestUtil.java:163]: This is in server-scm which doesn't use block count. Here existing test case is just to determine whether replica has some keys or not. If some keys are present then setEmpty as false.

**Comment by sodonnel** [comment]
This change looks mostly good. I have added one comment about a potential compatibility issue we might have, but it should be easily fixed.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java
Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java]: Thanks for the review @sodonnel, I have handled this scenario now.

**Review by sodonnel — APPROVED** [review]
LGTM provided we get a green CI run.

**Comment by sodonnel** [comment]
Somehow github has only run the title check, and not all the others. Could you push an empty commit to this PR trigger the jobs again please?

**Comment by ashishkumar50** [comment]
> Somehow github has only run the title check, and not all the others. Could you push an empty commit to this PR trigger the jobs again please?

Pushed empty commit.

**Comment by siddhantsangwan** [comment]
I triggered CI, it was asking for approval.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java:91]: Why remove the check for number of keys being zero?

Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java:91]: This is the intention of change, We don't want to depend on keyCount(which can be non zero even though container is empty) to determine whether container is empty or not, in some situation keyCount has gone wrong. Instead now we use empty flag for checking whether replica is empty or not based on the results from all replica, and that will judge container is empty or not.

Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java:91]: That makes sense but then we're not checking whether SCM expects a container to be empty on the basis of `ContainerInfo` (0 keys currently) against what the replicas are actually reporting. For example, consider an edge case where some replicas are empty and some are non empty because of arbitrary reasons. If the non empty replicas are lost, then `EmptyContainerHandler` will now mark this container as `EMPTY` on the basis of reports from other replicas. Whereas we probably want to call the container `MISSING`?

Comment (sodonnel) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java:91]: This is a good point. Do we need to have an isEmpty flag on the containerInfo. This would be set to true when all the replicas report as empty. Right now, we are depending on the keyCount reported in the replicas, and when the count drops to zero in all replicas, then we update the keycount in containerInfo to zero. Perhaps we should update an empty flag in a similar way, when non of the replicas are reporting "isEmpty=false".

Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java:91]: > Do we need to have an isEmpty flag on the containerInfo. This would be set to true when all the replicas report as empty.

I agree with this suggestion.

Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java:91]: @siddhantsangwan, @sodonnel Thanks for pointing out. I have added flag for container info as well to determine whether all the replicas are empty. Can you please look again.

**Review by sodonnel — APPROVED** [review]
I think this looks good now. Would be good for @siddhantsangwan to have another check too.

**Review by siddhantsangwan — COMMENTED** [review]
Looks good to me overall. I think there can be some visibility issues between the writing thread of the BlockDeletingService writing to `isEmpty` and other threads reading this variable. They might not see the updated value.

**Comment by ashishkumar50** [comment]
> 

This will be used by container report, which will read eventually in same time or in the next interval which doesn't have any impact. Now also blockcount becomes 0 and it goes in container report eventually by other thread.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:407]: This is yet a third `isEmpty` method. We don't want any errors when checking if a container is empty, so there should be one clear method to call that handles all the checks.

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto
Comment (errose28) [inline hadoop-hdds/interface-client/src/main/proto/hdds.proto:240]: Similar comment as above, I don't think we should persist this value as part of the main container state in SCM. And it *definitely* should not default to true.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java
Comment (errose28) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java:402]: This value should be populated on DN startup when it is loading the containers into memory. There should not be a case where this value is unset since we do not support rolling upgrades with mixed versions yet.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java
Comment (errose28) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java:91]: > Do we need to have an isEmpty flag on the containerInfo. This would be set to true when all the replicas report as empty.

I don't think we should have this. We should only track each piece of information in one place. We track if all the replicas are empty based on the `isEmpty` flag of each replica. Now there would be a duplicate value that may not match the state of all the current replicas. The safest thing is to check the isEmpty flag on all replicas before deleting, making this extra flag redundant.

> For example, consider an edge case where some replicas are empty and some are non empty because of arbitrary reasons. If the non empty replicas are lost, then EmptyContainerHandler will now mark this container as EMPTY on the basis of reports from other replicas.

Having another `areAllEmpty` flag on the `ContainerInfo` does not change this case, because that flag is still set based on all the replicas that SCM sees. It is no different from checking all replicas at the time of container delete.

**Review by errose28 — COMMENTED** [review]
Thanks for taking this up @ashishkumar50. I have some comments on the delete flag handling on SCM. Also the following cases need test coverage:
- isEmpty flag is correctly set on a container when it is imported as part of replication.
- isEmpty flag is correctly set on a container after loading it from memory (restart)
- SCM will only send delete command when isEmpty is true on all replicas, regardless of block count or other metadata.

**Comment by ashishkumar50** [comment]
> Thanks for taking this up @ashishkumar50. I have some comments on the delete flag handling on SCM. Also the following cases need test coverage:
> 
> * isEmpty flag is correctly set on a container when it is imported as part of replication.
> * isEmpty flag is correctly set on a container after loading it from memory (restart)
> * SCM will only send delete command when isEmpty is true on all replicas, regardless of block count or other metadata.

Hi @errose28, Thanks for the review. All review comments are addressed and test cases added. Can you please look again.

**Comment by adoroszlai** [comment]
@errose28 if you are happy with the most recent changes, please hit the "Approve and run" button, too.  [CI passed](https://github.com/ashishkumar50/ozone/actions/runs/5009952551) in @ashishkumar50's fork.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java
Comment (sodonnel) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java]: What about when we do support rolling upgrades? Someone may be on a version where the DNs don't have the new field and then jump to a future version that does support rolling upgrades. IMO it is safer to try to be compatible, rather than saying "we don't support that" as that support may change. There have been quite a few places in the code where I have seen this sort of compatibility code, and I think it starts a dangerous precedent if we are not saying we don't need to do it.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java
Comment (sodonnel) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java:365]: I'm not sure we should have removed this change, as I think we should try to be compatible between mixed versions, as we don't know when rolling upgrades may happen, or what someone will try for upgrading.

Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java:365]: Agree, If rolling upgrade is planned now or in future this can impact. I think we can handle this change now. This change is an extra precaution and doesn't impact anything else.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java
Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java:390]: @errose28 @sodonnel, I have kept this change for now. So can we go ahead we this?

**Comment by sodonnel** [comment]
@errose28 Do you want to take another look at this? Most of the last comments were from yourself, so I'd like to give you a chance to check them before we commit. This PR has been open a long time so it would be good to get it done.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1301]: Should we change the name of this method so it is not confused with `KeyValueContainerData#isEmpty`? Maybe changing this method to `hasBlocks` or something like that?

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/TestKeyValueContainer.java
Comment (errose28) [inline hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/TestKeyValueContainer.java:824]: Is this step necessary? Shouldn't this be handled by `KeyValueContainer#importContainerData` like in `testEmptyContainerImportExport`?

**Review by errose28 — COMMENTED** [review]
Thanks for the updates @ashishkumar50 and sorry for my delay getting back to this. I do have some more inline comments though.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java
Comment (errose28) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java:407]: Compatibility guarantees have to be strictly defined and tested. For example, we support non-rolling upgrades from all versions since 1.0.0 and non-roling downgrades pre-finalization to all versions since 1.1.0. We have tests for this and require changes to adhere to these rules. The compatibility code you are referring to is to comply to these guidelines.

Rolling upgrade has not been planned yet. When it is, we will define an earliest version that supports rolling upgrade, tests, and requirements that developers must follow, just like we have for non-rolling upgrade. Without this, enforcement of rolling upgrade compatibility will be arbitrary. We cannot attempt to support the feature in this state. There are also issues that guarantee no code getting checked in to master right now will be eligible for rolling upgrade support:

- OM HA commit model is incompatible with rolling upgrades. There may be other areas of the code as well that currently prevent us from doing rolling upgrades from this version, but we have not done an invovled investigation yet.

- There are no test suites to catch any rolling upgrade incompatibilities that are introduced, so it is impossible to guarantee the current version would work with a rolling upgrade. If it happend to work it would be purely by coincidence.

- There is no agreement among developers on rolling upgrade version guarantees so requirements are not being enforced in code reviews.

Once the rolling upgrade feature is proposed and requirements are defined then we will have explicit rules to guide changes and versions. Without this framework in place, enforcing rolling upgrade requirements in PRs just leads to confusing code, because it is documenting and designing for a situation that cannot happen. For that reason I propose removing the extra check, even though it does not affect correctness in the current version.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/TestKeyValueContainer.java
Comment (ashishkumar50) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/TestKeyValueContainer.java:824]: We need that you can look in `KeyValueHandler#importContainer`
we first call `populateContainerPathFields(...)`
and then `container.importContainerData(...)`
Even the other test(`testContainerImportExport`) has done the same.

In `testEmptyContainerImportExport` we are creating new container, and then reusing the same which was already initialised in `createContainer()`

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java
Comment (ashishkumar50) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java]: Updated to `hasBlocks`

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestDeleteContainerHandler.java
Comment (ashishkumar50) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestDeleteContainerHandler.java:220]: Added updated metric.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestBlockDeletion.java
Comment (ashishkumar50) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestBlockDeletion.java:660]: Added high time interval for RM.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestBlockDeletion.java
Comment (ashishkumar50) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestBlockDeletion.java]: Updated to use `TestHelper#waitForContainerClose`

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestDeleteContainerHandler.java
Comment (ashishkumar50) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestDeleteContainerHandler.java:318]: Since we have already created key before and not deleted, Key will be in block table. So again writing into rocks db block table may not be required. We are just deleting blocks from the disk here. In this case block table has entry but disk don't have, So in this case container will not be empty and delete is not allowed unless we do force delete.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestLegacyReplicationManager.java
Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestLegacyReplicationManager.java]: Added check in `isContainerEmpty()` to also verify replica should not be empty. Restored this test case.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java
Comment (sumitagrawl) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1331]: checkIfNoBlockFiles is unused, so can remove member variable and initialization

**Review by sumitagrawl — COMMENTED** [review]
@ashishkumar50 Thanks for working over this, have a minor comment

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java
Comment (ashishkumar50) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:1331]: Removed, Thanks.

**Review by sumitagrawl — APPROVED** [review]
@ashishkumar50 Thanks for working over this, LGTM +1

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java
Comment (sodonnel) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:207]: Today I came across a scenario:

1. Client tries to write an EC block, but at roughly the same time a node is stopped it was about to write to.
2. The client tries to write out the stripe to the data and parity nodes, but if fails to goto all the nodes as one node is down. The client notices this and abandons the stripe and writes the data again to a new container.
3. Now we have a case were some containers have a chunk written, but no "put block" got executed so RocksDB has no entry to the chunk / blocks.
4. The container is reported to SCM and closed (trigged by the stale node handler closing the pipeline) with zero keys and zero bytes (due to the missing put block).
5. Goes into delete handing in SCM, which sends a delete command over and over, as it is skipped by the DN due to the container not being empty.

Feels like the existing check is too conservative, as this is a scenario that is valid to happen, and we need to be able to remove these replicas.

I am wondering if the changes here would help or would the behavior be the same, leaving the container stuck in the DELETING state forever?

**Comment by sodonnel** [comment]
Where are we with getting this committed? Its been open a long time now.

Also, I am concerned the checks on the DN to remove a container are too conservative. If RocksDB has no entry for a block, I am not sure we should by default check for chunks on disk - with no RocksDB data we cannot access them anyway and the checksum data will be lost (it is stored in RocksDB).

See my comment earlier today and also more detail in https://issues.apache.org/jira/browse/HDDS-8838. This is an example of a container that should be deleted, but it will not delete (even before this change, but I don't think this change will alter that behavior).

I also wonder if the default state for a containerData object should be "empty" and then it gets toggled to "non empty" on the first successful put block to it - that may be one way to catch this scenario, but also it is possible for a container to have an orphan block like this, and other blocks. The other blocks could get removed over time, leaving the container empty except for this. orphan block.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java
Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:207]: We can resolve this by defaulting `hdds.datanode.check.empty.container.delete` to false. That might better in a separate PR. We could also change the config name to specify it only controls checking the directory. RocksDB will always be checked.

**Comment by errose28** [comment]
> I also wonder if the default state for a containerData object should be "empty" and then it gets toggled to "non empty" on the first successful put block to it - that may be one way to catch this scenario, 

This would complicate SCM code. Instead of sending delete commands to containers with the empty flag set, SCM would have to figure out if the container is empty because it needs to be deleted, or empty because it has not accepted any writes yet. I guess the allocated state could be used to make this distinction, but it would add another complication on the delete path. It also does not resolve your other point below.

> it is possible for a container to have an orphan block like this, and other blocks. The other blocks could get removed over time, leaving the container empty except for this. orphan block.

As I mentioned above, I think changing the default value of the config to check the directory on container delete is the best solution here.

**Review by errose28 — COMMENTED** [review]
LGTM. We still need to resolve [this thread](https://github.com/apache/ozone/pull/4655#discussion_r1195824586) on upgrades before deciding to merge though.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java
Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java:407]: Since rolling upgrade is not yet supported, Once it starts supporting that will be first version and future code needs to handle accordingly. So before support there will not be recommendation to do rolling upgrade. 
@sodonnel @errose28, Please confirm shall I remove this compatibility change?

Comment (sodonnel) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java]: OK - it is fine to remove the compatibility part. I am sure I have seen (and had review comments against my own code) code in the past that does this compatibility thing, which I why I raised it, but if we really don't need to worry about it for now, then that is good to know for the future.

**Comment by ashishkumar50** [comment]
> LGTM. We still need to resolve [this thread](https://github.com/apache/ozone/pull/4655#discussion_r1195824586) on upgrades before deciding to merge though.

@errose28, This conversation has been resolved. As currently we don't support rolling upgrade, incompatible code fix has been removed.

**Review by errose28 — APPROVED** [review]
Thanks for the continuous updates @ashishkumar50. LGTM. I triggered CI and we should be good to merge when it passes.

Can you do a follow up PR to change the default value of the directory check to false, leaving only the RocksDB check? It should hopefully be a small change.

**Comment by ashishkumar50** [comment]
> Thanks for the continuous updates @ashishkumar50. LGTM. I triggered CI and we should be good to merge when it passes.
> 
> Can you do a follow up PR to change the default value of the directory check to false, leaving only the RocksDB check? It should hopefully be a small change.

Ok @errose28, I will create a Jira and raise PR for that.

**Pull request merged**
